src/Domain/Services/CheckoutFields.php

```diff
@@ -0,0 +1,653 @@
+		}
+
+		$field = $this->additional_fields[ $key ];
+		if ( ! empty( $field['required'] ) && empty( $value ) ) {
+			// translators: %s field key.
+			return new \WP_Error( 'woocommerce_blocks_checkout_field_required', \sprintf( __( 'The field %s is required.', 'woo-gutenberg-products-block' ), $key ) );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Returns true if the given key is a valid field.
+	 *
+	 * @param string $key The field key.
+	 *
+	 * @return bool True if the field is valid, false otherwise.
+	 */
+	public function is_field( $key ) {
+		return array_key_exists( $key, $this->additional_fields );
+	}
+
+	/**
+	 * Persists a field value for a given order. This would also optionally set the field value on the customer.
+	 *
+	 * @param string    $key The field key.
+	 * @param mixed     $value The field value.
+	 * @param \WC_Order $order The order to persist the field for.
+	 * @param bool      $set_customer Whether to set the field value on the customer or not.
+	 *
+	 * @return void
+	 */
+	public function persist_field_for_order( $key, $value, $order, $set_customer = true ) {
+		$this->set_array_meta( $key, $value, $order );
+		if ( $set_customer ) {
+			if ( isset( wc()->customer ) ) {
+				$this->set_array_meta( $key, $value, wc()->customer );
+			} elseif ( $order->get_customer_id() ) {
+				$customer = new \WC_Customer( $order->get_customer_id() );
+				$this->set_array_meta( $key, $value, $customer );
+			}
+		}
+	}
+
+	/**
+	 * Persists a field value for a given customer.
+	 *
+	 * @param string       $key The field key.
+	 * @param mixed        $value The field value.
+	 * @param \WC_Customer $customer The customer to persist the field for.
+	 *
+	 * @return void
+	 */
+	public function persist_field_for_customer( $key, $value, $customer ) {
+		$this->set_array_meta( $key, $value, $customer );
+	}
+
+	/**
+	 * Sets a field value in an array meta, supporting routing things to billing, shipping, or additional fields, based on a prefix for the key.
+	 *
+	 * @param string                 $key The field key.
+	 * @param mixed                  $value The field value.
+	 * @param \WC_Customer|\WC_Order $object The object to set the field value for.
+	 *
+	 * @return void
+	 */
+	private function set_array_meta( $key, $value, $object ) {
+		$meta_key = '';
+
+		if ( 0 === strpos( $key, '/billing/' ) ) {
+			$meta_key = self::BILLING_FIELDS_KEY;
+			$key      = str_replace( '/billing/', '', $key );
+		} elseif ( 0 === strpos( $key, '/shipping/' ) ) {
+			$meta_key = self::SHIPPING_FIELDS_KEY;
+			$key      = str_replace( '/shipping/', '', $key );
+		} else {
+			$meta_key = self::ADDITIONAL_FIELDS_KEY;
+		}
+
+		if ( $object instanceof \WC_Customer ) {
+			if ( ! $object->get_id() ) {
+				$meta_data = wc()->session->get( $meta_key, array() );
+			} else {
+				$meta_data = get_user_meta( $object->get_id(), $meta_key, true );
+			}
```

Comment:
If we fetch the data from user data if we have a logged in customer, otherwise we get it from session.